### ISSUE TITLE
On Linux, "audio device selector" toolbars is too small and dropdowns are cut off

### ISSUE THREAD

**Issue by nyanpasu64** [tenacityteam/tenacity-legacy#435]
- [x] I have read the specified guidelines for issues

**Describe the bug**
On Linux, the "audio device selector" toolbar is too small and cuts off the left side of the combo boxes/dropdowns.

**To Reproduce**
1. Install and run Tenacity (or Audacity) on Linux.

**Expected behavior**
The toolbars are expanded, or the dropdowns are shrunken, so the audio device selector dropdowns don't have their left sides cut off.

As a workaround, you can drag the right side of the toolbar to expand it until the dropdowns are fully visible.

**Screenshots**

https://twitter.com/TenacityAudio/status/1424146535386865668

![screenshot](https://pbs.twimg.com/media/E8OVfydXIAc3Agy?format=jpg&name=4096x4096)

**Additional information (please complete the following information):**
 - OS: I'm using Arch Linux and KDE.
 - Version: long-standing issue since Audacity 2.4.x.

**Additional context**

Looking at the tweet linked above, apparently the issue also happens on GNOME. I think this issue happens on all Linux DEs where WxWidgets is backed by wxGTK and GTK 3, on both low and high DPI desktops, but I don't encounter it on Windows.

**Comment by Be-ing** [comment]
The screenshot on the right is how it looks with wxWidgets 3.1.5 on a high DPI screen. I don't know if this affects lower resolution screens. I also do not know if the bug is in Tenacity or wxWidgets.

**Comment by CHJ85** [comment]
Maybe a pop-up dialog would be a better solution than the drop down menu?

**Comment by Be-ing** [comment]
Why is this even in the main window?

**Comment by akleja** [comment]
Channel assignment should be on tcp, channel count on track creation, audio device / host in device settings imo.

Also the strange decision to have field description on the two first lines of the channel count selector.
And being able to select how many channels, but not which of those channels, which makes recording on a multi channel interface through ALSA virtually impossible.

**Comment by Be-ing** [comment]
Mockups or PRs would be great @akleja! I don't have the capacity to follow the design discussions in detail and work on the build system simultaneously but you seem to have good ideas.

**Comment by akleja** [comment]
I do have a mockup in the works. So far it's limited to the main window, but I plan on extending it to include some rework of the settings. The tcp needs a lot of love as well. 

Regarding the toolbar in question, all those settings are present in the device settings in the prefs window. Removing it would not hurt in the least imo. But maybe replacing it with a small status toolbar that shows buffer size and samplerate and which acts as a shortcut to the device settings might be a good replacement. I'll see what I can come up with.

Thank you! You're doing an amazing work on the build system.

**Comment by akleja** [comment]
Perhaps an alternative toolbar could look something like this?
![device_bar](https://user-images.githubusercontent.com/49785979/128658494-062962d0-605e-457e-ab7f-96d7f5ae69a7.png)
Just a quick initial mockup. The cog would turn red when the audio device is offline.

**Comment by Be-ing** [comment]
> replacing it with a small status toolbar that shows buffer size and samplerate and which acts as a shortcut to the device settings might be a good replacement.

Good idea.

**Comment by Be-ing** [comment]
Wait, are the audio buffer size and sample rate really so important to always be visible on screen in the main window? Maybe just the cog would be sufficient.

**Comment by akleja** [comment]
No perhaps not, that might just be my personal preferences. Perhaps those could be shown in the corner of the status bar?
But just removing the toolbar and adding a shortcut to the settings might not be enough.
I've heard a lot of people (mostly non musicians) switch between multiple devices like headsets and such often. I'll have to give this more thought.

**Comment by akleja** [comment]
I've been thinking about the issue about how to be able to change audio device with ease.
Here's a mockup of one solution, using the icons of the input/output level sliders as buttons, and using wxMenu to select device.
![device_2](https://user-images.githubusercontent.com/49785979/129421480-4a966ade-f4b6-45c6-b037-89de7360328c.png)
I haven't  begun working on implementing it yet, wanted feedback first. But after looking at the code it should be fairly straight forward.
Another idea would be to merge the sliders to sit under their corresponding meter, and use the icon in a similar way. Might be less confusing and less scattered.

I have begun moving some stuff to the device settings, so things aren't scattered all over the place. Default sample rate was no problem, but moving the current/project sample rate is a bit more involved since it needs quite a bit of restructuring work to tie it free from the project selection manager and put it back into the project manager. Also I'm not super sure if we need both settings, but I guess they both serve their purposes.

Assigning input channels, and selecting which track to record on is a related issue, but that I haven't looked into yet. I'd prefer to set that on the TCP. Right now you can't even select input channel, and the behavior of what track things get recorded on is super strange to me.

**Comment by Be-ing** [comment]
> using the icons of the input/output level sliders as buttons, and using wxMenu to select device.
> Another idea would be to merge the sliders to sit under their corresponding meter, and use the icon in a similar way. Might be less confusing and less scattered.

I like both of these ideas.

**Comment by akleja** [comment]
Cool, I'll look into it a bit more after the weekend. I'm leaning towards merging sliders and meters. It'll be a bigger challenge, and will take some time, but it's probably worth it in the end.

**Comment by Be-ing** [comment]
I hope you don't mean overlapping the sliders and the meters like Traktor:
https://cdn.digitaldjtips.com/app/uploads/2010/10/09145942/traktor-pro1.jpg
That obscures the view of the meters. If you mean putting the sliders adjacent to the meters, yeah, good idea.

**Comment by akleja** [comment]
Oh yeah no way. I agree that'd be utterly useless.
I'd thinking meter above, slider bellow, and menu button to the left.
I'll whip up a mockup later before starting with implementation just to be extra clear.

**Comment by akleja** [comment]
I've started moving things around and transforming the meter toolbars to input/output toolbars.
Here's a screenshot of how it's looking so far:
![io_toolbar](https://user-images.githubusercontent.com/49785979/129560185-18a84ae4-a6af-47c2-ba2d-db88bc8b682d.png)
The next step is adding device selection to the menus, which will be the bigger challenge.
And then improving the resizing/placement of elements in the toolbars. Right now I haven't implemented a vertical layout, but I've at least planned for it.

**Comment by Be-ing** [comment]
That looks good, keep going!

**Comment by akleja** [comment]
I've been making some progress. The functionality is there at last. The sizing works ok, though it needs some more work.
The layout now flips when resized as it should.
There's a lot left to be done, but I'm getting there (slowly)
![io_tb_3](https://user-images.githubusercontent.com/49785979/129815830-aeea8702-777f-4c71-bef6-25a4a820aecb.png)

**Comment by Be-ing** [comment]
Nice. Could you open a draft pull request?

**Comment by akleja** [comment]
Yeah I'll have to do a bit of cleanup first, but I'll open one when that's done.

**Comment by akleja** [comment]
Another question. Do we really need the device menu in the "Extra" menu? Currently it's tied into the combo boxes in the device toolbar. I guess I could migrate the code into a more suitable place. It just seems a bit redundant with the device settings existing and all, but perhaps it serves a purpose?

**Comment by Be-ing** [comment]
Where is the "Extra" menu that you are referring to?

**Comment by akleja** [comment]
It has to be enabled in prefs/interface

**Comment by Be-ing** [comment]
Yeah, that Extras > Device menu should be removed. It doesn't even do anything on my machine. Also the whole Extra menu should probably be broken up and moved to appropriate places. IMO "extras" and "advanced" modes are always cop-outs for actually making a good design.

**Comment by akleja** [comment]
Yeah I agree. I'll remove the Extras > Device menu for now. The rest has to be sorted out later, but a lot of it seems superfluous to me.

**Comment by akleja** [comment]
I just tried using a screen reader on linux. The screen reader support isn't that great. I'm thinking it might be a good idea to add the new i/o menus to the Extras menu anyway for now. Maybe later all the stuff in the extras menu can be moved out into the regular menus, but not having the device options in the menus is probably a bad idea for visually impaired users who would have a hard time using the toolbars. And at least with Orca, menus are a lot smoother to navigate than the prefs window.
